kernel/memory/Vmm.cpp

```diff
@@ -74,27 +172,51 @@ namespace Npk::Memory
         sl::ScopedLock scopeLock(rangesLock);
 
         globalLowerBound = PageSize; //dont allocate in page 0.
-        globalUpperBound = ~hhdmBase;
+        globalUpperBound = ~hhdmBase - PageSize;
         hatMap = InitNewMap();
 
+        for (size_t i = 0; i < (size_t)VmmMetaType::Count; i++)
+            metaSlabs[i] = nullptr;
+
+        VmHole* initialHole = new(AllocMeta(VmmMetaType::Hole)) VmHole();
+        initialHole->base = globalLowerBound;
+        initialHole->length = globalUpperBound - globalLowerBound;
+        initialHole->largestHole = initialHole->length;
+        holes.Insert(initialHole);
+
         Log("User VMM initialized.", LogLevel::Info);
     }
 
     VMM::VirtualMemoryManager(VmmKey)
     {
         sl::ScopedLock scopeLock(rangesLock);
 
-        //protect hhdm from allocations, we also reserve a region right after hhdm for early slab allocation.
+        //protect the HHDM from allocations, as well as the kernel binary.
         globalLowerBound = hhdmBase + hhdmLength;
-        globalUpperBound = sl::AlignDown((uintptr_t)KERNEL_BLOB_BEGIN, GiB);
+        globalUpperBound = sl::AlignDown((uintptr_t)KERNEL_BLOB_BEGIN, GetHatLimits().modes[0].granularity);
+
+        for (size_t i = 0; i < (size_t)VmmMetaType::Count; i++)
+            metaSlabs[i] = nullptr;
+
+        VmHole* initialHole = new(AllocMeta(VmmMetaType::Hole)) VmHole();
```

Comment:
i see this code repeated twice, can you make it a private method or something? Or is there a subtle difference that i can't see ?